File: sdk/inferlet/python/src/inferlet/eta/builder.py
"""
The DSL channel and the neutral trace builder — ports of
`eta-dsl/src/channel.rs` and `eta-dsl/src/builder.rs` (+ `lint.rs`).

`DslChannel` is the trace-side half of a channel (declaration + endpoint
claims). `Builder` takes descriptor-port bindings and stage closures, traces
the closures once into the canonical `TraceContainer`, derives every
channel's host role, and runs the span lints. It knows nothing of WIT; the
`bridge` module wraps it.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable

from .ir import (
    ChannelDecl,
    HostRole,
    Port,
    PortBinding,
    SinkScope,
    Stage,
    StageProgram,
    TraceContainer,
    container_hash,
    numel,
    shape_of,
)
from .trace import (
    ChannelState,
    SinkCall,
    StageResult,
    TraceError,
    intern_channel,
    is_tracing,
    next_gid,
    record_channel_put,
    record_channel_read,
    trace_stage,
    with_constants,
    with_session,
)
from .value import ConstData, Tensor, const_data, materialize, reshape_id_to


class DslChannel:
    """A handle to a channel's trace state. Inside a traced stage,
    `take`/`read`/`put` record the IR's channel ops."""

    __slots__ = ("state",)

    def __init__(self, state: ChannelState) -> None:
        self.state = state

    @staticmethod
    def _build(shape, dtype, capacity: int, seed: ConstData | None, seeded: bool) -> "DslChannel":
        gid = next_gid()
        state = ChannelState(
            gid=gid,
            name=f"ch{gid}",
            shape=shape_of(shape),
            dtype=dtype,
            capacity=capacity,
            seeded=seeded or seed is not None,
            has_seed=seed is not None,
        )
        return DslChannel(state)

    @staticmethod
    def new(shape, dtype) -> "DslChannel":
        return DslChannel._build(shape, dtype, 1, None, False)

    @staticmethod
    def from_const(data: ConstData) -> "DslChannel":
        return DslChannel._build(data.shape, data.dtype, 1, data, True)

    @staticmethod
    def seeded(shape, dtype) -> "DslChannel":
        return DslChannel._build(shape, dtype, 1, None, True)

    # -- declaration ------------------------------------------------------

    def capacity(self, n: int) -> "DslChannel":
        self.state.capacity = n
        return self

    def named(self, name: str) -> "DslChannel":
        self.state.name = name
        return self

    @property
    def gid(self) -> int:
        return self.state.gid

    @property
    def name(self) -> str:
        return self.state.name

    @property
    def shape(self):
        return self.state.shape

    @property
    def dtype(self):
        return self.state.dtype

    @property
    def is_seeded(self) -> bool:
        return self.state.seeded

    # -- endpoint claims ----------------------------------------------------

    def note_host_put(self) -> None:
        self.state.host_puts += 1

    def note_host_take(self) -> None:
        self.state.host_takes += 1

    def note_host_read(self) -> None:
        self.state.host_reads += 1

    def note_desc_claim(self, consumes: bool) -> None:
        if consumes:
            self.state.desc_takes += 1
        else:
            self.state.desc_reads += 1

    # -- in-program ops ------------------------------------------------------

    def take(self) -> Tensor:
        if not is_tracing():
            raise TraceError(
                f"channel {self.name}: take() outside a stage body is a host operation; "
                "use `await ch.take_host()`"
            )
        vid, ty = record_channel_read(self.state, True)
        return Tensor.node(vid, ty)

    def read(self) -> Tensor:
        if not is_tracing():
            raise TraceError(
                f"channel {self.name}: read() outside a stage body is a host operation; "
                "use `await ch.read_host()`"
            )
        vid, ty = record_channel_read(self.state, False)
        return Tensor.node(vid, ty)

    def put_tensor(self, t: Tensor) -> None:
        if not is_tracing():
            raise TraceError(f"channel {self.name}: put(Tensor) outside a traced stage")
        vid, ty = materialize(t)
        fitted = reshape_id_to(vid, ty, self.state.shape)
        record_channel_put(self.state, fitted)


# ---------------------------------------------------------------------------
# Builder
# ---------------------------------------------------------------------------


@dataclass
class Traced:
    """A traced, linted forward pass: the canonical container plus the
    channels it references, in container (dense) order."""

    container: TraceContainer
    channels: list[ChannelState]

    def encode(self) -> bytes:
        return self.container.encode()

    def identity_hash(self) -> int:
        return container_hash(self.encode())


class Builder:
    def __init__(self, vocab: int, page_size: int) -> None:
        self.ports: list[tuple[Port, DslChannel]] = []
        self.stages: list[tuple[Stage, Callable[[], None]]] = []
        self.vocab = vocab
        self.page_size = page_size

    def bind_port(self, port: Port, source: DslChannel) -> None:
        """Bind a descriptor port to a channel, recording the port's endpoint
        claim per its consumption discipline."""
        source.note_desc_claim(port.consumes)
        self.ports.append((port, source))

    def stage(self, stage: Stage, body: Callable[[], None]) -> None:
        for i, (s, _) in enumerate(self.stages):
            if s == stage:
                self.stages[i] = (stage, body)
                return
        self.stages.append((stage, body))

    def _channel_port(self, port: Port) -> DslChannel | None:
        for p, ch in self.ports:
            if p == port:
                return ch
        return None

    def _rows(self) -> int:
        ch = self._channel_port(Port.READOUT)
        if ch is not None:
            return max(min(numel(ch.shape), 0xFFFF_FFFF), 1)
        ch = self._channel_port(Port.EMBED_INDPTR)
        if ch is not None:
            return max(min(numel(ch.shape), 0xFFFF_FFFF) - 1, 1)
        return 1

    def _record(self, rows: int) -> tuple[list[StageResult], list[tuple[Port, int]]]:
        ports: list[tuple[Port, int]] = []
        for port, source in self.ports:
            ports.append((port, intern_channel(source.state)))
        results: list[StageResult] = []
        for stage in Stage:
            body = None
            for s, b in self.stages:
                if s == stage:
                    body = b
            if body is None:
                continue
            results.append(trace_stage(stage, rows, body))
        return results, ports

    def build(self) -> Traced:
        rows = self._rows()
        (stage_results, ports), channels, names = with_constants(
            self.vocab, self.page_size, lambda: with_session(lambda: self._record(rows))
        )

        # Re-key the container to gid (declaration) order.
        order = sorted(range(len(channels)), key=lambda i: channels[i].gid)
        remap = [0] * len(channels)
        for new_idx, old_idx in enumerate(order):
            remap[old_idx] = new_idx
        channels = [channels[i] for i in order]
        for r in stage_results:
            for op in r.ops:
                if op.chan >= 0:
                    op.chan = remap[op.chan]
        ports = [(p, remap[ci]) for p, ci in ports]

        # Name table: strictly sorted and unique.
        name_order = sorted(range(len(names)), key=lambda i: names[i])
        name_remap = [0] * len(names)
        for new_idx, old_idx in enumerate(name_order):
            name_remap[old_idx] = new_idx
        names = [names[i] for i in name_order]
        for r in stage_results:
            for op in r.ops:
                if op.tag in (0xA1, 0xA2):
                    op.name_idx = name_remap[op.name_idx]

        sinks: list[tuple[Stage, SinkCall]] = [(r.stage, s) for r in stage_results for s in r.sinks]

        decls: list[ChannelDecl] = []
        for st in channels:
            has_prog_put = bool(st.prog_puts)
            has_prog_consume = bool(st.prog_takes) or bool(st.prog_reads)
            has_desc_use = st.desc_takes > 0 or st.desc_reads > 0
            has_host_put = st.host_puts > 0
            host_consumes = st.host_takes > 0 or st.host_reads > 0
            is_terminal_output = (
                has_prog_put
                and not has_prog_consume
                and not has_desc_use
                and not has_host_put
                and not st.seeded
                and not st.has_seed
            )
            # A seeded channel the pass only reads (a descriptor, or a program
            # `read` such as a control word) is a latest-value cell replaceable
            # through host `set`, so it needs a Writer endpoint too.
            seeded_latest_value_writer = st.seeded and (has_desc_use or bool(st.prog_reads)) and not has_prog_put
            if (has_host_put or seeded_latest_value_writer) and not has_prog_put:
                host_role = HostRole.WRITER
            elif host_consumes and (bool(st.prog_takes) or has_prog_put):
                host_role = HostRole.READER
            elif is_terminal_output:
                host_role = HostRole.READER
            else:
                host_role = HostRole.NONE
            seeded = st.seeded or (has_host_put and has_prog_put)
            decls.append(ChannelDecl(st.shape, st.dtype, st.capacity, host_role, seeded))

        stages = [StageProgram(r.stage, r.ops) for r in stage_results]
        port_bindings = [PortBinding(p, ci) for p, ci in sorted(ports, key=lambda x: int(x[0]))]
        container = TraceContainer(names=names, channels=decls, ports=port_bindings, stages=stages)

        _lint(channels, sinks)

        return Traced(container=container, channels=channels)


def _lint(channels: list[ChannelState], sinks: list[tuple[Stage, SinkCall]]) -> None:
    errs: list[str] = []
    for st in channels:
        host_writes = st.host_puts > 0
        host_consumes = st.host_takes > 0 or st.host_reads > 0
        stage_puts = bool(st.prog_puts)
        stage_consumes = bool(st.prog_takes) or st.desc_takes > 0
        if host_writes and host_consumes:
            errs.append(
                f"channel `{st.name}` has two host endpoints (host writes and host consumes); "
                "SPSC needs one pass endpoint"
            )
        produced = stage_puts or host_writes or st.seeded or st.has_seed
        consumed = stage_consumes or bool(st.prog_reads) or st.desc_reads > 0 or host_consumes
        if consumed and not produced:
            errs.append(f"channel `{st.name}` is consumed but never produced or seeded")
    for stage, s in sinks:
        if s.scope == SinkScope.PASS_WIDE:
            ok = stage == Stage.PROLOGUE
        else:
            ok = stage in (Stage.PROLOGUE, Stage.ON_ATTN_PROJ)
        if not ok:
            errs.append(f"sink `{s.name}` is misplaced in stage `{stage.wire_name}`")
    if errs:
        raise TraceError("; ".join(errs))

File: sdk/inferlet/javascript/src/eta/builder.ts
// The DSL channel and the neutral trace builder — ports of
// `eta-dsl/src/channel.rs` and `eta-dsl/src/builder.rs` (+ `lint.rs`).

import {
  ChannelDecl,
  Dtype,
  HostRole,
  Port,
  PortBinding,
  SinkScope,
  Stage,
  STAGE_NAMES,
  StageProgram,
  TraceContainer,
  containerHash,
  encode,
  numel,
  portConsumes,
  shapeOf,
  tags,
} from './ir.js';
import {
  ChannelState,
  SinkCall,
  StageResult,
  TraceError,
  internChannel,
  isTracing,
  nextGid,
  recordChannelPut,
  recordChannelRead,
  traceStage,
  withConstants,
  withSession,
} from './trace.js';
import { ConstData, Tensor, materialize, reshapeIdTo } from './value.js';

/** A handle to a channel's trace state. */
export class DslChannel {
  constructor(readonly state: ChannelState) {}

  private static build(shape: readonly number[], dtype: Dtype, capacity: number, seed: ConstData | null, seeded: boolean): DslChannel {
    const gid = nextGid();
    const state: ChannelState = {
      gid,
      name: `ch${gid}`,
      shape: shapeOf(shape),
      dtype,
      capacity,
      seeded: seeded || seed !== null,
      hasSeed: seed !== null,
      progPuts: [],
      progTakes: [],
      progReads: [],
      hostPuts: 0,
      hostTakes: 0,
      hostReads: 0,
      descTakes: 0,
      descReads: 0,
      host: null,
    };
    return new DslChannel(state);
  }

  static new(shape: readonly number[], dtype: Dtype): DslChannel {
    return DslChannel.build(shape, dtype, 1, null, false);
  }
  static fromConst(data: ConstData): DslChannel {
    return DslChannel.build(data.shape, data.dtype, 1, data, true);
  }
  static seeded(shape: readonly number[], dtype: Dtype): DslChannel {
    return DslChannel.build(shape, dtype, 1, null, true);
  }

  capacity(n: number): this {
    this.state.capacity = n;
    return this;
  }
  named(name: string): this {
    this.state.name = name;
    return this;
  }
  get gid(): number {
    return this.state.gid;
  }
  get name(): string {
    return this.state.name;
  }
  get shape() {
    return this.state.shape;
  }
  get dtype(): Dtype {
    return this.state.dtype;
  }
  get isSeeded(): boolean {
    return this.state.seeded;
  }

  noteHostPut(): void {
    this.state.hostPuts++;
  }
  noteHostTake(): void {
    this.state.hostTakes++;
  }
  noteHostRead(): void {
    this.state.hostReads++;
  }
  noteDescClaim(consumes: boolean): void {
    if (consumes) this.state.descTakes++;
    else this.state.descReads++;
  }

  take(): Tensor {
    if (!isTracing()) throw new TraceError(`channel ${this.name}: take() outside a stage body is a host operation; use ch.takeHost()`);
    const [vid, ty] = recordChannelRead(this.state, true);
    return Tensor.node(vid, ty);
  }
  read(): Tensor {
    if (!isTracing()) throw new TraceError(`channel ${this.name}: read() outside a stage body is a host operation; use ch.readHost()`);
    const [vid, ty] = recordChannelRead(this.state, false);
    return Tensor.node(vid, ty);
  }
  putTensor(t: Tensor): void {
    if (!isTracing()) throw new TraceError(`channel ${this.name}: put(Tensor) outside a traced stage`);
    const [vid, ty] = materialize(t);
    const fitted = reshapeIdTo(vid, ty, this.state.shape);
    recordChannelPut(this.state, fitted);
  }
}

/** A traced, linted forward pass: the canonical container plus the
 * channels it references, in container (dense) order. */
export class Traced {
  constructor(
    readonly container: TraceContainer,
    readonly channels: ChannelState[],
  ) {}

  encode(): Uint8Array {
    return encode(this.container);
  }
  identityHash(): bigint {
    return containerHash(this.encode());
  }
}

export class Builder {
  private ports: [Port, DslChannel][] = [];
  private stages: [Stage, () => void][] = [];

  constructor(
    readonly vocab: number,
    readonly pageSize: number,
  ) {}

  bindPort(port: Port, source: DslChannel): void {
    source.noteDescClaim(portConsumes(port));
    this.ports.push([port, source]);
  }

  stage(stage: Stage, body: () => void): void {
    const i = this.stages.findIndex(([s]) => s === stage);
    if (i >= 0) this.stages[i] = [stage, body];
    else this.stages.push([stage, body]);
  }

  private channelPort(port: Port): DslChannel | undefined {
    return this.ports.find(([p]) => p === port)?.[1];
  }

  private rows(): number {
    const ro = this.channelPort(Port.READOUT);
    if (ro) return Math.max(Math.min(numel(ro.shape), 0xffff_ffff), 1);
    const ei = this.channelPort(Port.EMBED_INDPTR);
    if (ei) return Math.max(Math.min(numel(ei.shape), 0xffff_ffff) - 1, 1);
    return 1;
  }

  private record(rows: number): { results: StageResult[]; ports: [Port, number][] } {
    const ports: [Port, number][] = this.ports.map(([port, source]) => [port, internChannel(source.state)]);
    const results: StageResult[] = [];
    for (const stage of [Stage.PROLOGUE, Stage.ON_ATTN_PROJ, Stage.ON_ATTN, Stage.EPILOGUE]) {
      let body: (() => void) | undefined;
      for (const [s, b] of this.stages) if (s === stage) body = b;
      if (!body) continue;
      results.push(traceStage(stage, rows, body));
    }
    return { results, ports };
  }

  build(): Traced {
    const rows = this.rows();
    const {
      result: { results: stageResults, ports: rawPorts },
      channels: rawChannels,
      names: rawNames,
    } = withConstants(this.vocab, this.pageSize, () => withSession(() => this.record(rows)));

    // Re-key the container to gid (declaration) order.
    const order = rawChannels.map((_, i) => i).sort((a, b) => rawChannels[a].gid - rawChannels[b].gid);
    const remap = new Array<number>(rawChannels.length);
    order.forEach((oldIdx, newIdx) => (remap[oldIdx] = newIdx));
    const channels = order.map((i) => rawChannels[i]);
    for (const r of stageResults) for (const op of r.ops) if (op.chan >= 0) op.chan = remap[op.chan];
    const ports: [Port, number][] = rawPorts.map(([p, ci]) => [p, remap[ci]]);

    // Name table: strictly sorted and unique.
    const nameOrder = rawNames.map((_, i) => i).sort((a, b) => (rawNames[a] < rawNames[b] ? -1 : rawNames[a] > rawNames[b] ? 1 : 0));
    const nameRemap = new Array<number>(rawNames.length);
    nameOrder.forEach((oldIdx, newIdx) => (nameRemap[oldIdx] = newIdx));
    const names = nameOrder.map((i) => rawNames[i]);
    for (const r of stageResults) for (const op of r.ops) if (op.tag === tags.KERNEL_CALL || op.tag === tags.SINK_CALL) op.nameIdx = nameRemap[op.nameIdx];

    const sinks: [Stage, SinkCall][] = [];
    for (const r of stageResults) for (const s of r.sinks) sinks.push([r.stage, s]);

    const decls: ChannelDecl[] = channels.map((st) => {
      const hasProgPut = st.progPuts.length > 0;
      const hasProgConsume = st.progTakes.length > 0 || st.progReads.length > 0;
      const hasDescUse = st.descTakes > 0 || st.descReads > 0;
      const hasHostPut = st.hostPuts > 0;
      const hostConsumes = st.hostTakes > 0 || st.hostReads > 0;
      const isTerminalOutput = hasProgPut && !hasProgConsume && !hasDescUse && !hasHostPut && !st.seeded && !st.hasSeed;
      // A seeded channel the pass only reads (a descriptor, or a program
      // `read` such as a control word) is a latest-value cell replaceable
      // through host `set`, so it needs a Writer endpoint too.
      const seededLatestValueWriter = st.seeded && (hasDescUse || st.progReads.length > 0) && !hasProgPut;
      let hostRole: HostRole;
      if ((hasHostPut || seededLatestValueWriter) && !hasProgPut) hostRole = HostRole.WRITER;
      else if (hostConsumes && (st.progTakes.length > 0 || hasProgPut)) hostRole = HostRole.READER;
      else if (isTerminalOutput) hostRole = HostRole.READER;
      else hostRole = HostRole.NONE;
      const seeded = st.seeded || (hasHostPut && hasProgPut);
      return { shape: st.shape, dtype: st.dtype, capacity: st.capacity, hostRole, seeded };
    });

    const stages: StageProgram[] = stageResults.map((r) => ({ stage: r.stage, ops: r.ops }));
    const portBindings: PortBinding[] = ports.sort((a, b) => a[0] - b[0]).map(([port, channel]) => ({ port, channel }));
    const container: TraceContainer = { names, channels: decls, ports: portBindings, stages, externs: [] };

    lint(channels, sinks);

    return new Traced(container, channels);
  }
}

function lint(channels: ChannelState[], sinks: [Stage, SinkCall][]): void {
  const errs: string[] = [];
  for (const st of channels) {
    const hostWrites = st.hostPuts > 0;
    const hostConsumes = st.hostTakes > 0 || st.hostReads > 0;
    const stagePuts = st.progPuts.length > 0;
    const stageConsumes = st.progTakes.length > 0 || st.descTakes > 0;
    if (hostWrites && hostConsumes) errs.push(`channel \`${st.name}\` has two host endpoints (host writes and host consumes); SPSC needs one pass endpoint`);
    const produced = stagePuts || hostWrites || st.seeded || st.hasSeed;
    const consumed = stageConsumes || st.progReads.length > 0 || st.descReads > 0 || hostConsumes;
    if (consumed && !produced) errs.push(`channel \`${st.name}\` is consumed but never produced or seeded`);
  }
  for (const [stage, s] of sinks) {
    const ok = s.scope === SinkScope.PASS_WIDE ? stage === Stage.PROLOGUE : stage === Stage.PROLOGUE || stage === Stage.ON_ATTN_PROJ;
    if (!ok) errs.push(`sink \`${s.name}\` is misplaced in stage \`${STAGE_NAMES[stage]}\``);
  }
  if (errs.length) throw new TraceError(errs.join('; '));
}
